### PR TITLE
Integrated reactotron into application with dev configuration and added console.logs()

### PR DESCRIPTION
**Pull request by HarshulSharma000**
Current integration of Reactotron is not tied to dev evironment. I have configured it in a way that it is only active for dev evironment and has no affect in production. I guess this one is the correct one. And I have moved to Reactotron v1 because that is current stable version of Reactotron. Although v2 is also out but it is in beta stage currently and is known to have issues with windows. So for use in Zulip I think we should consider V1 for now.
For docs on how and why I did it this way please refer to:-
https://github.com/infinitered/reactotron/blob/v1.x/docs/quick-start-react-native.md
https://github.com/infinitered/reactotron/blob/v1.x/docs/plugin-redux.md
https://github.com/infinitered/reactotron/blob/v1.x/docs/tips.md#running-in-production

### CODE REVIEW

File: index.android.js [hunk 1]
```diff
@@ -1,5 +1,5 @@
 /* @flow */
-// import './src/ReactotronConfig';
+import './src/ReactotronConfig';
```
Comment (HarshulSharma000) [inline index.android.js]: Include Reactotron into our execution flow and it must be done at first line as per docs.

File: package.json [hunk 2]
```diff
@@ -108,8 +108,8 @@
     "jest-react-native": "^18.0.0",
     "prettier": "^1.11.1",
     "react-native-cli": "^2.0.1",
-    "reactotron-react-native": "^2.0.0-beta.1",
-    "reactotron-redux": "^2.0.0-beta.1",
+    "reactotron-react-native": "^1.14.0",
+    "reactotron-redux": "^1.13.0",
```
Comment (HarshulSharma000) [inline package.json]: Reverting to previous more stable version of reactotron.

File: src/ReactotronConfig.js [hunk 3]
```diff
@@ -5,16 +5,19 @@ import { NativeModules } from 'react-native';
...
 
-Reactotron.configure(/*{ host: scriptHostname }*/) // controls connection & communication settings
+if (__DEV__) {
+  const scriptURL = NativeModules.SourceCode.scriptURL;
+  const scriptHostname = scriptURL.split('://')[1].split(':')[0]; //To retrieve scriptURL from device
+
+  console.logs = (text) => Reactotron.log(text,true);
+
+  //Comment out next lines to disable reactotron connection
+  Reactotron.configure({ host: scriptHostname, name: "Zulip Mobile" }) // controls connection & communication settings
```
Comment (HarshulSharma000) [inline src/ReactotronConfig.js]: Surrounded whole code within `__Dev__` so that there is no harm in production environment.

File: src/boot/store.js [hunk 4]
```diff
@@ -2,20 +2,21 @@
...
-// uncomment the following lines to integrate reactotron with redux
-// const store = Reactotron.createStore(
-//   rootReducer,
-//   compose(autoRehydrate(), applyMiddleware(...middleware)),
-// );
+let store; 
 
-const store = compose(applyMiddleware(...middleware), autoRehydrate())(createStore)(rootReducer);
+if(__DEV__) {
+  store = Reactotron.createStore(rootReducer, {},compose(applyMiddleware(...middleware), autoRehydrate()))
```
Comment (HarshulSharma000) [inline src/boot/store.js]: Since Reactotron creates a lightweight wrapper around our store, its no harm switching to normal store creation in production environment althought it is not really necessary because reactotron.createStrore() does not add much of an overhead.

File: src/ReactotronConfig.js [hunk 5]
```diff
@@ -5,16 +5,19 @@ import { NativeModules } from 'react-native';
...
 
 // uncomment the following lines to test on a device
-// let scriptHostname;
-// if (__DEV__) {
-//   const scriptURL = NativeModules.SourceCode.scriptURL;
-//   scriptHostname = scriptURL.split('://')[1].split(':')[0];
-// }
 
-Reactotron.configure(/*{ host: scriptHostname }*/) // controls connection & communication settings
+if (__DEV__) {
```
Comment (HarshulSharma000) [inline src/ReactotronConfig.js]: Surrounded whole Reactotron related code in `__DEV__` so that nothing else gets executed beyond dev environment.

File: src/ReactotronConfig.js [hunk 6]
```diff
@@ -1,20 +1,21 @@
 /* eslint-disable */
-import Reactotron, { asyncStorage, networking, openInEditor } from 'reactotron-react-native';
-import { NativeModules } from 'react-native';
+if (__DEV__) {
```
Comment (HarshulSharma000) [inline src/ReactotronConfig.js]: Moved all the Required code into `__DEV__` flag so that none of the Reactotron related code gets executed in the production build. This makes zero impact on production build.

File: src/ReactotronConfig.js [hunk 7]
```diff
@@ -1,20 +1,21 @@
...
-// uncomment the following lines to test on a device
-// let scriptHostname;
-// if (__DEV__) {
-//   const scriptURL = NativeModules.SourceCode.scriptURL;
-//   scriptHostname = scriptURL.split('://')[1].split(':')[0];
-// }
-
-Reactotron.configure(/*{ host: scriptHostname }*/) // controls connection & communication settings
+  //Comment out next lines to disable reactotron connection
+  Reactotron.configure({ host: scriptHostname, name: "Zulip Mobile" }) // controls connection & communication settings
```
Comment (HarshulSharma000) [inline src/ReactotronConfig.js]: Just added the name of our app that appears in Reactotron Timeline.

File: src/boot/store.js [hunk 8]
```diff
@@ -27,4 +38,6 @@ export const restore = (onFinished?: () => void) =>
     onFinished,
   );
 
-export default store;
+const exportStore = store; // Just a hack to avoid eslint error of exporting non const value.
```
Comment (HarshulSharma000) [inline src/boot/store.js]: This was not required but Eslint was too fixated that let declared variables must not be imported. So its just a hack to satisfy eslint and is not at all required.

File: src/boot/store.js [hunk 9]
```diff
@@ -9,13 +9,24 @@ import middleware from './middleware';
...
 // AsyncStorage.clear(); // use to reset storage during development
 
-// uncomment the following lines to integrate reactotron with redux
-// const store = Reactotron.createStore(
-//   rootReducer,
-//   compose(autoRehydrate(), applyMiddleware(...middleware)),
-// );
+let store;
 
-const store = compose(applyMiddleware(...middleware), autoRehydrate())(createStore)(rootReducer);
```
Comment (HarshulSharma000) [inline src/boot/store.js]: Moved Reactotron store creation in `__DEV__` block so that normal store is created in production environment.

File: src/boot/store.js [hunk 10]
```diff
@@ -9,13 +9,24 @@ import middleware from './middleware';
...
+if (__DEV__) { // This block of code is never meant to execute in production
```
Comment (HarshulSharma000) [inline src/boot/store.js]: Moved Reactotron store creation in `__DEV__` block so that normal store is created in production environment.

**Comment by gnprice** [comment]
Hi @HarshulSharma000 !

The build is failing with lint errors -- you need to fix the build before we can merge the PR. If you google for the errors, you'll quickly find some good answers for how to fix them appropriately.

Can you explain what `console.logs` is about?

File: index.android.js (hunk 1 above)
Comment (borisyankov) [reply index.android.js]: I suspect the start of `ZulipMobile.js` is good enough and you do not need to repeat it in two files.

File: package.json (hunk 2 above)
Comment (borisyankov) [reply package.json]: What issues did you experience?

**Comment by HarshulSharma000** [comment]
@gnprice It's just a log function used to make logs on reactotron desktop app. I consider it a very lightweight addition to standard console and in any way it does not affect our production.

File: package.json (hunk 2 above)
Comment (HarshulSharma000) [reply package.json]: Some of my api request's body were not getting logged. Also the order of actions being fired was changed from the original order. Those little pesky quirks are still there So i think we should wait for v2 to release in stable version and then migrate.

File: index.android.js (hunk 1 above)
Comment (HarshulSharma000) [reply index.android.js]: Ok I will do that.

File: .gitignore [hunk 11]
```diff
@@ -64,3 +64,9 @@ package-lock.json
 # https://guides.cocoapods.org/using/using-cocoapods.html#should-i-check-the-pods-directory-into-source-control
 #
 ios/Pods/
+
+#Generated files
+#
+generated/
+intermediates/
+outputs/
```
Comment (borisyankov) [inline .gitignore]: Are these changes related to Reactotron?

Comment (HarshulSharma000) [reply .gitignore]: Outputs should not be included in git repo I did that on my side while setting up the project initially.  Should I remove them? Create a separate PR? or let them be?

Comment (borisyankov) [reply .gitignore]: Do not include anything that is not related to this PR.

Comment (borisyankov) [reply .gitignore]: If you think it is useful for all other contributors create another PR and explain when these folders are generated. Optionally you can add these to your local file and tell `git` to ignore the changes to this file.

**Comment by borisyankov** [comment]
Elaborate on this please:
> And I think the failures are necessary for the very essence of this PR.

**Comment by HarshulSharma000** [comment]
@borisyankov    
The errors which are being encountered right now are mostly linting and are as follows:-
`14:5  error '__DEV__' is not defined no-undef  ` - `__DEV__` is automatically defined by react native env so js has no knowlege about it. But `__DEV__`'s usage is considered best for detecting development environment.

`15:22  error Unexpected require() global-require` - Requires are not hoisted so they are definitely needed to make things dynamic otherwise Reactotron will be imported regardless of the need. They are required to import things conditionally. Even the Reactotron docs suggest that. [https://github.com/infinitered/reactotron/blob/v1.x/docs/tips.md#running-in-production](Running in production).
`15:22  error 'reactotron-react-native' should be listed in the project's dependencies, not devDependencies import/no-extraneous-dependencies`- It is added as dev dependencies because we don't want it to be added to production by mistake( and also is a redundant package for production environment). That's why even the previous person who added Reactotron before me did the same.

File: .gitignore (hunk 11 above)
Comment (HarshulSharma000) [reply .gitignore]: OK I am making it a new PR.

File: src/ZulipMobile.js [hunk 12]
```diff
@@ -1,4 +1,5 @@
 /* @flow */
+import './ReactotronConfig';
```
Comment (HarshulSharma000) [inline src/ZulipMobile.js]: Reactotron is needed to be imported in first line of your project if not done so sometimes due to changes in ordering of imports it results in errors.
Refrences:-
Docs- https://github.com/infinitered/reactotron/blob/v1.x/docs/quick-start-react-native.md 
Checkout importing reactotron portion
Commmon Errors- https://github.com/infinitered/reactotron/issues/289 
P.S. - I did spend a day or two with the same so I guess its better for all of us to keep it that way.

**Comment by gnprice** [comment]
@HarshulSharma000 I think both Boris and I can read the build failure just fine :-), and we can both see that the failures are entirely in lint.

A basic fact of our workflow (and really any sane development workflow) is: we do not merge changes that cause the build to fail. (Or if we ever do by mistake, we promptly fix it.) That's the only way the build stays useful.

If you think a lint error should be ignored, the remedy is to tell the build to ignore it, so the build stops failing. In that case, also explain why in a comment and/or the commit message.

Again, if you google for these lint errors, you can quickly find some good answers for how to silence them appropriately.

**Comment by HarshulSharma000** [comment]
@gnprice How about I disable those specific rules for the files that necessarily require it? That way I won't affect the overall config and eslint won't throw those warnings either.

**Comment by gnprice** [comment]
(Done now in #3109).
